---
title: "Phylogenetic Tree Construction"
author: "Andrea Martinez"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---
# Goals of this file

The goal is to create a phylogentic tree!

1. Load in preprocessed phyloseq object.
2. Create ASV file from the phyloseq object.
3. Align the 16S sequencing from the fasta file with MAFFT.
4. Create a tree with FastTree2.

## Set the seed
```{r set-seed}
set.seed(238428)
```


## Load Packages & Phyloseq Object
```{r load-environment}
# phytools, ggtree, RColorBrewer
pacman::p_load(phytools, ggtree, RColorBrewer, install = FALSE)

# Load physeq
load("data/02_PreProcessing/raw_preprocessed_phyloseq.RData")
raw_preprocessed_phyloseq
```

# Create Fasta file of ASV and their Sequences

This fasta file will be used to create our alignment in MAFFT

```{r create-preprocessed-fast}
# pull out ASV seqs and ASV names
ASVseq_df<-raw_preprocessed_phyloseq@tax_table %>%
  data.frame() %>%
  dplyr::select(ASV, ASVseq)

#View(ASVseq_df)
# Add the > to make fasta header
ASVseq_df$ASV <- paste0(">", ASVseq_df$ASV)
#View(ASVseq_df)
# Create the fasta object
ASVseq_fasta<-c(rbind(ASVseq_df$ASV,ASVseq_df$ASVseq))
head (ASVseq_fasta)
# Write to file
write(ASVseq_fasta,
      file = "data/03_PhylogeneticTree/preprocessed_ASV.fasta")
```

## Align the 16S sequences from fasta file with MAFFT

- `engine.opts  = '-l'`: gives us original terminal base environment 

```{r run-mafft, engine = 'bash', engine.opts = '-l'}
#Write bash code to run mafft
# First provide the path to mafft
export PATH=/programs/mafft/bin:$PATH

#Change directories to provide the fasta file we made above
cd data/03_PhylogeneticTree/
pwd

# Set a seed - use saem seed as before for consistency
RANDOM=238428

# Run Mafft
# To test in the shell directly from rmd
  # For Mac: command + option + enter
# For now, use default options, not the version
# MAFFT automaically knows that it's a nucleotide alignment

/programs/mafft/bin/mafft --auto preprocessed_ASV.fasta > MAFFT_aligned_ASVs.fasta

# Change back to the project directory
cd ../../
pwd
```

# FastTree2
```{r run-fasttree, engine = 'bash', engine.opts = '-l'}
# Provide export path to fasttree
export PATH=/programs/FastTree-2.1.11:$PATH

# cd into the alignment file filder
cd data/03_PhylogeneticTree
pwd

# Run Fasttree to generate phylogenetic tree
# parameters: 
  # -nt (indicates nucleotide alignment)
  # =gtr = generalized time-reversible nucleotide substitution model (9 parameters)
  # -fastest = speed up the model, reduce memory usage (recommended for datasets that have > 50,000 seqs)
  # -log = output a log file
  # input alignment file
  # specify the output tree file
  
FastTree -nt -gtr -fastest -log FastTree.log MAFFT_aligned_ASVs.fasta > ASVs_unrooted.tree

# Change back to the main directory
cd ../../
pwd
echo "The working directory is $PWD"
```
# Session Information
```{r session-info}
devtools::session_info()
```
